src/gui/editors/PianoRoll.cpp

```diff
@@ -712,6 +712,85 @@ void PianoRoll::glueNotes()
 	}
 }
 
+void PianoRoll::fitNoteLengths(bool fill)
+{
+	if (!hasValidPattern()) { return; }
+	m_pattern->addJournalCheckPoint();
+
+	// Reference notes
+	NoteVector refNotes = m_pattern->notes();
+	std::sort(refNotes.begin(), refNotes.end(), Note::lessThan);
+
+	// Notes to edit
+	NoteVector notes = getSelectedNotes();
+	if (notes.empty())
+	{
+		notes = refNotes;
+	}
+	else if (!fill)
+	{
+		std::sort(notes.begin(), notes.end(), Note::lessThan);
+	}
+	if (fill)
+	{
+		std::sort(notes.begin(), notes.end(), [](Note *n1, Note *n2) { return n1->endPos() < n2->endPos(); });
+	}
+
+	int length;
+	NoteVector::iterator ref = refNotes.begin();
+	for (Note *note : notes)
+	{
+		// Fast forward to next reference note
+		while (ref != refNotes.end() && (fill ? (*ref)->pos() < note->endPos() : (*ref)->pos() <= note->pos()))
+		{
+			ref++;
+		}
+		if (ref == refNotes.end())
+		{
+			// Last notes stretch to end of last bar
+			length = notes.last()->endPos().nextFullBar() * TimePos::ticksPerBar() - note->pos();
```

Comment:
```suggestion
			// Last notes stretch to end of last bar
			length = note->endPos().nextFullBar() * TimePos::ticksPerBar() - note->pos();
```

I'm still trying to think if there's a possibility that `ref` reaches `refNotes.end()` before `note` reaches `notes.last()`. But that change doesn't create an issue if it doesn't, but prevents an issue if it does.

Reply:
I think it's the desired behavior in fill mode. In cut mode, this `if` clause is causing a bug:
1. Create a 1/2 bar note
2. Create another 3/2 bars note at the same position and higher key
3. `Cut overlaps` may truncate the 3/2 bars note to 1 bar(I'm not sure if the behavior is well-defined, but at least Linux+GCC builds allow reproducing this bug)

Reply:
I think `length` should be `note->length()` in cut mode.

Reply:
My bad, for some reason I thought `notes.last()` was the one being resized, but it was only the one being used as a reference to get the last bar. It should work fine for the fill action then (using `note->endPos()` instead would fill every note that doesn't have a following reference note to the next bar, which might be different for each one. Not sure if that behavior would be desired).

But it ended up making you find that bug you mentioned, exactly because `notes.last()` is being used. Since having the 3/2 bars note in a higher key makes the 1/2 bar note the last, its next full bar is used instead.

Reply:
I wouldn't mind `length` being used instead. But for the user, wouldn't it be preferred that notes that "overlap" the last bar are cut to fit in?

Reply:
For the sake of simplicity, maybe for now we should go with the `length` solution. The idea of cutting the chord to the last bar would require us to go through all notes of the chord to see which is shortest before calculating the length.

That should do it then:
```cpp
length = fill
? note.last()->endPos().nextFullBar() * TimePos::ticksPerBar() - note->pos()
: note->length()
```

Reply:
Good find, both of you. I don't think Cut should cut notes where the last bar line is, that makes little sense. I'll just add

```c++
if (!fill) { continue; }
```

Reply:
It's good by me, better than my suggestion!